htlcswitch/htlcnotifier/htlcnotifier.go

```diff
@@ -0,0 +1,221 @@
+	// If it failed on the outgoing link, it is false.
+	IncomingFailed bool
+
+	// Timestamp is the time when the link failure occurred.
+	Timestamp time.Time
+}
+
+// ForwardingFailEvent represents a HTLC failure which occurred down the line
+// after we forwarded the HTLC through our node. An error is not included in
+// this event because errors returned down the route are encrypted.
+type ForwardingFailEvent struct {
+	// HTLCKey contains the information which uniquely identifies the original
+	// HTLC forwarded event.
+	HTLCKey
+
+	// Timestamp is the time when the forwarding failure was received.
+	Timestamp time.Time
+}
+
+// SettleEvent represents a HTLC that was forwarded by our node and successfully
+// settled.
+type SettleEvent struct {
+	// HTLCKey contains the information which uniquely identifies the original
+	// HTLC forwarded event.
+	HTLCKey
+
+	// HTLCInfo contains further information about the HTLC.
+	HTLCInfo
+
+	// Timestamp is the time when this HTLC was settled.
+	Timestamp time.Time
+}
+
+// NotifyForwardingEvent notifies the HTLCNotifier than a HTLC has been
+// forwarded.
+func (h *HTLCNotifier) NotifyForwardingEvent(event ForwardingEvent) {
+	log.Debugf("Notifying forward event from %v(%v) to %v(%v)",
+		event.IncomingChannel, event.HTLCInIndex, event.OutgoingChannel,
+		event.HTLCOutIndex)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send forwarding event: %v", err)
+	}
+}
+
+// NotifyLinkFailEvent notifies the HTLCNotifier that we have failed a HTLC on
+// one of our links.
+func (h *HTLCNotifier) NotifyLinkFailEvent(event LinkFailEvent) {
+	log.Debugf("Notifying link failure event from %v(%v) to %v(%v),"+
+		"incoming link: %v, with reason: %v, details: %v", event.IncomingChannel,
+		event.HTLCInIndex, event.OutgoingChannel, event.HTLCOutIndex,
+		event.IncomingFailed, event.FailReason, event.FailDetail)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send link fail event: %v", err)
+	}
+}
+
+// NotifyForwardingFailEvent notifies the HTLCNotifier that a HTLC we forwarded
+// has failed down the line.
+func (h *HTLCNotifier) NotifyForwardingFailEvent(event ForwardingFailEvent) {
+	log.Debugf("Notifying forwarding failure event from %v(%v) to %v(%v)",
+		event.IncomingChannel, event.HTLCInIndex, event.OutgoingChannel,
+		event.HTLCOutIndex)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send forwarding fail event: %v", err)
+	}
+}
+
+// NotifySettleEvent notifies the HTLCNotifier that a HTLC that we committed to
+// as part of a forward or a receive to our node has been settled.
+func (h *HTLCNotifier) NotifySettleEvent(event SettleEvent) {
+	log.Debugf("Notifying settle event from %v(%v) to %v(%v)",
+		event.IncomingChannel, event.HTLCInIndex, event.OutgoingChannel,
+		event.HTLCOutIndex)
+
+	if err := h.ntfnServer.SendUpdate(event); err != nil {
+		log.Warnf("Unable to send settle event: %v", err)
+	}
+}
```

Comment:
Question that comes up is which event (sequences) to expect in the different scenarios:
* Incoming exit hop htlc rejected
* Incoming exit hop htlc settled
* Incoming exit hop htlc accepted but not yet resolved (hodl invoice), then settled
* Incoming exit hop htlc accepted but not yet resolved (hodl invoice), then rejected
* Outgoing origin htlc (our payment) not handed out (didn't even leave the link)
* Outgoing origin htlc (our payment) failed back by peer
* Outgoing origin htlc (our payment) settled
* Incoming htlc to forward not accepted
* Incoming htlc to forward accepted, but rejected by outgoing link
* Htlc forwarded, but failed back by downstream peer
* Htlc forwarded and settled

Are you sure that we are able to cover and distinguish all we need with the four events defined above? Are in any of the scenarios above multiple events sent?